modules/src/core/ics26_routing/context.rs

```diff
@@ -17,4 +34,188 @@ pub trait Ics26Context:
     + PortReader
     + Ics20Context
 {
+    type Router: Router;
+
+    fn router(&self) -> &Self::Router;
+
+    fn router_mut(&mut self) -> &mut Self::Router;
+}
+
+#[derive(Debug, PartialEq)]
+pub struct InvalidModuleId;
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct ModuleId(String);
+
+impl ModuleId {
+    pub fn new(s: Cow<'_, str>) -> Result<Self, InvalidModuleId> {
+        if !s.trim().is_empty() && s.chars().all(char::is_alphanumeric) {
+            Ok(Self(s.into_owned()))
+        } else {
+            Err(InvalidModuleId)
+        }
+    }
+}
+
+impl fmt::Display for ModuleId {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl FromStr for ModuleId {
+    type Err = InvalidModuleId;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Self::new(Cow::Borrowed(s))
+    }
+}
+
+impl Borrow<str> for ModuleId {
+    fn borrow(&self) -> &str {
+        self.0.as_str()
+    }
+}
+
+pub trait Acknowledgement: AsRef<[u8]> {
+    fn success(&self) -> bool;
+}
+
+pub type WriteFn = dyn FnOnce(&mut dyn Any);
+
+pub type DeferredWriteResult<T> = (Option<Box<T>>, Option<Box<WriteFn>>);
+
+// FIXME(hu55a1n1): Define concrete type that implements `Into<AbciEvent>`?
```

Comment:
> Why was the current implementation chosen over this suggestion?

This is because `ModuleEvent(tendermint::abci::Event)` is still Tendermint specific, and we need something that is chain agnostic, so it's still not clear to me what this type should be. We can also choose to go with something like an `IbcEvent` variant which is just a `BTreeMap<String, Vec<u8>>`, the benefit here is that this design prevents modules from being able to emit core Ibc events ->
```rust
type ModuleEvent = BTreeMap<String, Vec<u8>>;

// modules/src/events.rs ->
pub enum IbcEvent {
    // ...
    Module(ModuleEvent),
}
```
I believe these things will become more clear as we start having more users/customers.

> Github issues are usually a better tool for that.

I agree, I'll open a follow-up issue to document all these FIXMEs. 👍

Reply:
How about we make `ModuleEvent` be an associated type of `Module`, renamed`Event`. This seems to be the correct way to do it, since 2 `Module`s could potentially have different event types (if I understand correctly). Then `Module` takes a generic type `E` representing the type that `Event` needs to be convertible to (similar to what your comment suggests). Something like:

```rust
pub trait Module<E>: ... {
    type Event: Into<E>;     // For Tendermint apps, E = AbciEvent
    ...
}
```
Then, our implementation would not be tied to Tendermint, and different modules could have different event types.

However, the only problem I see is that currently, we have `TryFrom<IbcEvent> for AbciEvent` instead of `From<IbcEvent> for AbciEvent` (this seems to be to cover for the 2 special events `IbcEvent::Empty` and `IbcEvent::ChainError` which I currently don't understand). 

Does this make sense?